### DISCUSSION TITLE
How to specify schema/dataset for persisting tables?

### DISCUSSION THREAD

**Discussion by dimitri-b** [erezsh/Preql#28] [Q&A]
`preql` can persist tables using `table` command, but it is not clear which schema/dataset this table would be added to, if connected to a database that supports/uses schemas.

**Comment by erezsh** [comment]
This behavior isn't implemented yet. Creating a table will create it in the default schema.

I can adopt the postgres syntax:

```javascript
table my_schema.my_table = expression
```

Would that be sufficient for your purposes?

Are you using schemas heavily in your flow? If so, I will try to add full support soon.

**Reply by dimitri-b** [reply]
I think the postgres syntax will work fine for specifying datasets.

**Reply by erezsh** [reply]
@dimitri-b You might be interested to know I implemented this syntax.

Right now it's in the `schemas` branch ([See here](https://github.com/erezsh/Preql/tree/schemas)). I will include it in the next release, after I run a few more tests. But I thought you might like to try it earlier than that.

It also contains a few unrelated bugfixes for BigQuery.

**Reply by dimitri-b** [reply]
I tried in BQ and SQLite and got the same error in both cases.

<img width="798" alt="2021-08-27_09-49-40" src="https://user-images.githubusercontent.com/4684795/131050422-d9591dc5-8c6d-4141-813a-f51d774e65b3.png">

**Reply by erezsh** [reply]
It seems like you're not using the `schemas` branch. Possibly you need to uninstall the official release, and then install the branch.

I just pushed another commit to to `schemas`. Now if you're using it, when you run Preql you should see:

```
Preql 0.2.4-schemas interactive prompt
```

P.S. it won't work with sqlite, because it doesn't support schemas. (unless you can think of a workaround?)

**Reply by dimitri-b** [reply]
Thank you, I made some progress.

Everything seems to work OK, but something is going on with the `id` field (not part of the source table, must be a system field).

```
Exception traceback:
DbQueryError: Exception when trying to execute SQL code:
    INSERT INTO `eCom_US`.`temp_cal`(`Fin_Year`, `Fin_Year_Label`) SELECT `Fin_Year`, `Fin_Year_Label` FROM (SELECT * FROM (SELECT `Fin_Year`, `Fin_Year_Label` FROM `eCom_US`.`calendar`) LIMIT 5 OFFSET 0)

Got error: 400 Required field id cannot be null
```

Also, if the source table has any `DATE` or similar types - I get an error that they can't be loaded into `STRING` fields.

**Reply by erezsh** [reply]
Yes, that is a bug with the BigQuery implementation. I think it's because it doesn't support auto-increment fields.

I'll try to fix it soon!

I will also check regarding DATE columns. If you can give me an example of such source table (its definition, data isn't necessary) that would help.

**Reply by erezsh** [reply]
P.S. for such bugs and errors, feel free to open an issue, if you like. It might be more convenient to discuss each one separately.

**Reply by erezsh** [reply]
@dimitri-b I just fixed the bug you mentioned. You shouldn't be getting this error anymore. (in `schemas` branch)

But actually, any chance you can help me with a really weird bug I'm getting? The code to reproduce it is:

```go
 >> table A { x: int }
table B { x: int }

(Query took 2.38 seconds)
 >> SQL(list[int], """SELECT `x` FROM `a`.`B`  """)
table  =0
(Query took 2.16 seconds)
 >> SQL(nulltype, """INSERT INTO `a`.`A`(`x`) SELECT `x` FROM `a`.`B`  """)
Exception traceback:
DbQueryError: Exception when trying to execute SQL code:
    INSERT INTO `a`.`A`(`x`) SELECT `x` FROM `a`.`B`

Got error: 400 GET https://bigquery.googleapis.com/bigquery/v2/projects/aeyeconsole/queries/2c98463c-6b29-4a06-bb88-13e7ecb72bc6?location=US&prettyPrint=false: Table not found: a.B at [1:42]

(job ID: 2c98463c-6b29-4a06-bb88-13e7ecb72bc6)

            -----Query Job SQL Follows-----

    |    .    |    .    |    .    |    .    |    .    |
   1:INSERT INTO `a`.`A`(`x`) SELECT `x` FROM `a`.`B`
    |    .    |    .    |    .    |    .    |    .    |
```

What's going on here? It makes no sense to me, and it only happens with BigQuery.

**Reply by dimitri-b** [reply]
It worked for me in REPL, but I had to:
a) specify datasets for created tables, e.g. `table a.A {x: int}`, and
b) add the `id` column, otherwise I got _id cannot be null_ error if there is any data in the source table

It seems that if dataset is not specified, BQ lands the table in a random dataset of its choice, if there are several datasets. In my case it was the first listed dataset in the connected project.

I'm not sure why in your case the SELECT statement works, but INSERT doesn't.

```shell
 >> table test_data.A {x: int}
(Query took 2.37 seconds)
 >> table test_data.B {x: int}
(Query took 2.25 seconds)
 >> SQL(nulltype, """INSERT INTO `test_data`.`A`(`x`) SELECT `x` FROM `test_data`.`B`  """)
null
(Query took 2.90 seconds)
 >>
```

**Reply by erezsh** [reply]
@dimitri-b Just wanted to let you know, there is now full support for BigQuery in Preql (in the `master` branch. I'll make a release soon).

That is, all the tests I wrote are now passing. You might still find bugs, but it's probably going to happen a lot less.

I also opted to using a temporary `_preql` dataset for tables that aren't associated with a dataset, instead of choosing the first on the list. That means I create and delete the `_preql` dataset. Let me know if you think that's a bad idea.

**Reply by dimitri-b** [reply]
Defaulting to `_preql` dataset is a good idea, the only suggestions I have are:
- some users may not have the right permissions to create datasets, so a descriptive error message would be needed to save frustration
- on a remote chance that there is already a dataset with that name/id - what would you do? Overwrite or just use it? Delete it on session end? Another option would be to name the dataset using UUID or similar, but I don't like the sound of it, and it is probably not worth it.

**Reply by erezsh** [reply]
I suppose I can rename it to `_preql_432492834903280` or something like that. It felt to me that `_preql` is probably already unique enough. But since I would be deleting it without ceremony, perhaps it's better to take a very cautious approach.

**Reply by dimitri-b** [reply]
I think `_preql_432492834903280` pattern is a good approach - it would be clear to anyone seeing it that this is a temporary dataset managed by Preql.

**Comment by dimitri-b** [comment]
In case of BigQuery - our typical use case is to query a table from one dataset and save the results in another. We use and specify schema/dataset all the time. We also switch between dev/prod datasets/tables, via templating the SQL files.
